api/main.go

```diff
@@ -152,3 +153,195 @@ func (d *Dispatcher) Write(ctx context.Context, vl *ValueList) error {
 	}
 	return nil
 }
+
+type configValueType int
+
+const (
+	configTypeString configValueType = iota
+	configTypeNumber
+	configTypeBoolean
+)
+
+func (cvt configValueType) String() string {
+	return [3]string{"String", "Number", "Boolean"}[cvt]
+}
+
+// ConfigValue may be either a string, float64 or boolean value.
+// This is the Go equivalent of the C type "oconfig_value_t".
+type ConfigValue struct {
+	typ configValueType
+	s   string
+	f   float64
+	b   bool
+}
+
+func (v ConfigValue) String() (string, bool) {
+	return v.s, v.typ == configTypeString
+}
+
+func (v ConfigValue) Number() (float64, bool) {
+	return v.f, v.typ == configTypeNumber
+}
+
+func (v ConfigValue) Boolean() (bool, bool) {
+	return v.b, v.typ == configTypeBoolean
+}
+
+// Untyped returns the specific value of ConfigValue without specifying its type, useful for functions like fmt.Printf
+// which can use variables with unknown types.
+func (v ConfigValue) Untyped() interface{} {
+	switch v.typ {
+	case configTypeString:
+		return v.s
+	case configTypeNumber:
+		return v.f
+	case configTypeBoolean:
+		return v.b
+	}
+	return nil
+}
+
+// Config represents one configuration block, which may contain other configuration blocks.
+type Config struct {
+	Key      string
+	Values   []ConfigValue
+	Children []Config
+}
+
+// Merge appends other's children to c's children.
+// Returns an error if Key or any Values differ.
+func (c *Config) Merge(other *Config) error {
+	panic("Not yet implemented")
+}
+
+// Unmarshal applies the configuration from a Config to an arbitrary struct.
+func (c *Config) Unmarshal(v interface{}) error {
+	// Sanity check value of the interface
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Ptr || rv.IsNil() {
+		return fmt.Errorf("can only unmarshal to a non-nil pointer") // TODO: better error message or nil if preferred
+	}
+
+	// If the target supports unmarshalling let it
+	if u, ok := v.(Unmarshaler); ok {
+		return u.UnmarshalConfig(v)
+	}
+
+	drv := rv.Elem() // get dereferenced value
+	drvk := drv.Kind()
+
+	// If config has child configs we can only unmarshal to a struct or slice of structs
+	if len(c.Children) > 0 {
+		if drvk != reflect.Struct && (drvk != reflect.Slice || drv.Elem().Kind() != reflect.Struct) {
+			return fmt.Errorf("cannot unmarshal a config with children except to a struct or slice of structs")
+		}
+	}
+
+	switch drvk {
+	case reflect.Invalid, reflect.Array, reflect.Chan, reflect.Func, reflect.Interface, reflect.Map, reflect.UnsafePointer:
+		return fmt.Errorf("cannot unmarshal into type %s", drv.Type())
+	case reflect.Struct:
+		// Unmarshal values from config
+		if err := storeStructConfigValues(c.Values, drv); err != nil {
+			return fmt.Errorf("while unmarshalling values into %s: %s", drv.Type(), err)
+		}
+		for i := range c.Children {
```

Comment:
The code appears to only use `i` to access `c.Children`. This is uncommon in Go. The usual way to iterate over a slice is:

```go
for _, child := range c.Children {
  // child is of type "Config"
```